src/Implement/RequestHandler.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright (c) 2021, 鍾淯丞, 周杰仕, 林仁鴻. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "RequestHandler.h"
+
+#include <netinet/in.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <string>
+
+std::unordered_map<std::string, APIHandler *> RequestHandler::apis;
+
+RequestHandler::RequestHandler() { this->read_size = 4096; }
+
+void RequestHandler::SetFd(const int &file_descriptor) {
+  this->client_fd = file_descriptor;
+}
+
+int RequestHandler::ReadHeader(std::string &header,
+                               std::string &msg_over_header) {
+  char input_buffer[this->read_size];
+  const std::string find_key = "Content-Length:";
+  const std::string end_of_header = "\r\n\r\n";
+  // keep read buffer until no request message
+  while (true) {
+    int msg_read = recv(this->client_fd, input_buffer, sizeof(input_buffer), 0);
+    // means end of request
+    if (msg_read <= 0) {
+      return -1;
+    }
+    // add read buffer to header
+    header += input_buffer;
+    // find \r\n\r\n to get end of header
+    size_t start_cut_pos = header.find(find_key, 0) + find_key.length() + 1;
+    size_t end_cut_pos = header.find(end_of_header, 0);
+    if (end_cut_pos != std::string::npos) {
+      try {
+        // convert content-length string to int
+        int client_content_length = std::stoi(
+            header.substr(start_cut_pos, end_cut_pos - start_cut_pos));
+        // it may occur that reading too many info from buffer, need to move the
+        // body part out of header message
+        msg_over_header = header.substr(
+            end_cut_pos + end_of_header.length(),
+            header.length() - end_cut_pos - end_of_header.length());
+        header = header.substr(0, end_cut_pos + end_of_header.length());
+        return client_content_length;
+      } catch (...) {
+        // TODO: call error
+      }
+    }
+  }
+}
+
+std::string RequestHandler::ReadBody(const int &content_length,
+                                     const std::string &body,
+                                     std::string &next_request_part) {
+  std::string result = "";
+  std::string buffer_result = "";
+  if (content_length > body.length()) {
+    // buffer size is decided by content-length
+    char input_buffer[content_length - body.length()];
+    recv(this->client_fd, input_buffer, sizeof(input_buffer), 0);
```

Comment:
`recv()` is not guaranteed to read `sizeof(input_buffer)` amount of data. You'll need to loop through it as well.